add-reviews

rovniipoc opened the pull request:
Реализован функционал "Отзывы"

rovniipoc committed:
Добавить дата-класс Review

rovniipoc committed:
Добавить интерфейс ReviewStorage

rovniipoc committed:
Добавить роумаппер для отзывов ReviewRowMapper

rovniipoc committed:
Добавить ReviewDbStorage, изменить ReviewRowMapper, изменить Review и сопутствующие изменения

Изменения Review:
- тип данных reviewId, userId, filmId изменен с Long на Integer;

Изменения ReviewRowMapper:
- добавлен статический метод makeReview;

rovniipoc committed:
Добавить ReviewService

rovniipoc committed:
Добавить ReviewController, изменить ReviewDbStorage

Изменения ReviewDbStorage:
- удалено присвоение значения count по умолчанию, т.к. значение по умолчанию присваивается в ReviewController;

rovniipoc committed:
Изменить data.sql, изменить schema.sql

Изменения data.sql:
- добавлена очистка таблицы reviews при старте;

Изменения schema.sql:
- добавлена таблица reviews;

rovniipoc committed:
Добавить логирование в ReviewController

rovniipoc committed:
Добавить логирование в ReviewController, изменить ReviewDbStorage

Изменения ReviewDbStorage:
- исправить метод getByParams();

rovniipoc committed:
Изменить ReviewService, изменить ReviewDbService, изменить ReviewDbStorage

Изменения ReviewService:
- добавлен методы checkUserExistById(), checkCorrectReviewParams();
- в метод checkCorrectReviewParams() добавлены проверки на пустой контент и пустой тип отзыва;

Изменения ReviewDbService:
- теперь нельзя обновить userId и filmId у отзыва (на основании тестов postman);

Изменения ReviewDbStorage:
- исправлен метод update (теперь он возвращает взятый из репозитория обновленный отзыв);

rovniipoc committed:
Реализовать функционал отзывов, изменить ReviewDbStorage, изменить Review

Изменения ReviewDbStorage:
- теперь через метод update() нельзя изменить кол-во лайков (на основании тестов postman);

Изменения Review:
- Добавлено значение по умолчанию для поля useful;

rovniipoc committed:
Добавить плагин checkstyle

rovniipoc committed:
Исправить по чекстайлу

rovniipoc committed:
Изменить логику лайков отзывов для выполнения тестов postman ветки add-reviews

Теперь если пользователь поставил лайк, затем поставил дизлайк, то лайк убирается автоматически (и наоборот). Один пользователь может поставить не больше одного лайка(дизлайка)

Изменения ReviewDbStorage:
- изменены методы по работе с лайками, добавлен новый приватный метод обновления поля useful на основании лайков.

Изменения data.sql, schema.sql:
- добавлена таблица REVIEW_LIKES_DISLIKES содержащая информацию о том кто какому отзыву поставил лайк или дизлайк.

rovniipoc committed:
Исправить schema.sql

Изменения schema.sql:
- для референсных ключей таблиц reviews и review_likes_dislikes добавлены команды "on delete cascade"

artyom-petukhov-yapr commented on src/main/resources/schema.sql:
```diff
@@ -43,3 +43,20 @@ CREATE TABLE IF NOT EXISTS likes (
+CREATE TABLE IF NOT EXISTS review_likes_dislikes (
+    review_id int NOT NULL REFERENCES reviews(review_id) on delete cascade,
+    user_id int NOT NULL REFERENCES users(user_id) on delete cascade,
+    is_like boolean NOT NULL,
```
про is_like, is_dislike:
обычно "дублирование" источника данных приводит к проблемам, т.к. защиты на уровне бд нет от случаев is_like=true и is_dislike=true (is_like=false и is_dislike=false), а как их воспринимать неясно
может заменить на что-то типа `is_positive` (true - лайк, false - дизлайк)? и в коде должно стать по-проще

IvchenkoElena replied:
Тут на твое усмотрение, замечание дельное, но возможно обходить его будет слишком заморочено

artyom-petukhov-yapr commented on src/main/java/ru/yandex/practicum/filmorate/storage/ReviewDbStorage.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void addLike(Integer reviewId, Integer userId) {
+//        String query = "UPDATE reviews SET useful = ? WHERE review_id = ?";
```
может удалить закомментированные куски кода? обычно их не любят на ревью)) я про финальное ревью проекта

IvchenkoElena replied:
Красоту можно будет потом навести перед сдачей всей ветки develop

artyom-petukhov-yapr reviewed (approved):
оставил комментарии. на твое усмотрение. approve вжал

rovniipoc committed:
Изменить хранение лайков отзывов

Изменения:
- в schema.sql в таблице review_likes_dislikes удален столбец is_dislike, теперь информация и о лайках и о дизлайках хранится в столбце is_like, значение может быть: TRUE = лайк, FALSE = дизлайк, NULL = отсутствие лайка/дизлайка
- в ReviewDbStorage внесены соответствующие изменения в методы работы с лайками/дизлайками к отзывам
